bluesky-social/social-app
# [🐴] Error recovery

This PR started out as specifically block handling, but turned out to be easier to refactor everything at once.

**So included here is:**

Handling for when the user on the other end of the line blocks me and I don't know yet, so I send a message. It checks for block errors, and handles those as `unrecoverable` pending errors, whereas a network error (which I now check for) is `recoverable`. It also invalidates profile queries used elsewhere for shadowed optimistic block state using an emitter. Not super proud of this but it works and I think it's clear.

Better handling for network errors. Now if combinations of errors occur, the errors won't shift around, and resuming from one error will attempt to resume from other applicable errors.

It now attempts to recover when returning from background.

I also cleaned up the error components to be a little less invasive, since hopefully we'll clean stuff up before the user has a change to interact with them.

**Not in this PR**
The case of a user blocking me, I get into a blocked state, and then they unblock me, is not handled here. I expect that to be a rare occurrence, and it's only partially recoverable via the firehose e.g. if the blocker messages me again. Otherwise we'd need to poll for block state, and I don't think that's necessary atm. A simple refresh or pop/push back to the chat will reset it back to zero and the users can message again.

**Testing:**
- turn off network and send message, load history, etc
- turn network back on and recover
  - via interaction
  - via background->foreground
- block yourself from test account and then try to send a message, see block state in header

![CleanShot 2024-05-16 at 12 41 01@2x](https://github.com/bluesky-social/social-app/assets/4732330/839ea1f1-cff2-4a68-8575-e7a6c2126ded)
![CleanShot 2024-05-16 at 12 41 11@2x](https://github.com/bluesky-social/social-app/assets/4732330/b33bf3cc-4211-4cf2-b214-9f5be4a30dd4)
